Helper-side aggregation job init/continue asynchronous processing (#564)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

subsequent request and response, carries the remaining messages exchanged during
preparation.

Leader to Helper HTTP communication can be performed asynchronously or
synchronously. In the asynchronous case, the Leader and Helper exchange HTTP
requests that can be handled quickly, and the Leader polls the Helper for
aggregation job readiness. In the synchronous case, the Helper processes the
aggregation job while holding the request open, reducing the number of requests

> that are exchanged. This allows a Helper implementation flexibility in choosing

Reviewer: "the number of HTTP requests"

subsequent request and response, carries the remaining messages exchanged during
preparation.

> Leader to Helper HTTP communication can be performed asynchronously or
> synchronously. In the asynchronous case, the Leader sends aggregation job HTTP
> requests to the Helper that can be handled quickly. In the synchronous case, the
> Helper processes the aggregation job while holding the request open, reducing
> the number of HTTP requests that are sent. This choice allows a Helper

Reviewer: I think we can define "synchronous" and "asynchronous" more clearly.
```suggestion
Normally the Helper process each step synchronously, meaning it computes each step of the computation before producing a response to the Leaders HTTP request. The Helper also has the option to process each step asynchronously. This means the Helper responds to the request immediately, without producing the next step of the computation. To continue, the Leader polls the Helper until it responds with the next steep. This choice allows a Helper
```

Author: SGTM. I wordsmithed a bit and moved it down in this section, so that we're not introducing concepts that haven't been formally introduced yet (i.e. `steps`).

* `prepare_inits`: the sequence of `PrepareInit` messages constructed in the
  previous step.

Finally, the Leader sends a request to
`PUT {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`. The
payload is set to `AggregationJobInitReq` and the media type is set to
"application/dap-aggregation-job-init-req".

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

> After receiving HTTP status 201 Created, the Leader proceeds to poll the

Reviewer 2: What the recommended way of polling here? Do we poll each job in batch sequentially until all jobs give expected response? This would be blocking operation on leader side and generates noisy request to the helper. It would be better to provide ability to query for completed jobs. This enables leader to process-and-fetch the completed jobs in a loop.

Author: Aggregation jobs are independent, so it shouldn't be necessary to process jobs sequentially. Rather the leader can/should eagerly send aggregate init/continue to the helper.

> This would be blocking operation on leader side

Indeed, but it shouldn't be much worse than blocking leader-side waiting for the aggregate init/continue request to finish.

I'd actually expect this grants leader implementations more flexibility in how they schedule aggregation job processing. If a leader-side worker thread is blocked on an open HTTP request, it can't set it aside to do other work. In the proposed model, the leader-side worker can poll once, see that the aggregation job isn't ready, move on to other work (e.g. polling other workers), and poll again later.

> generates noisy request to the helper

Indeed that's a tradeoff, but I'm hoping that empty GET requests are fairly cheap in terms of bandwidth. And aggregator pairings are free to tune polling intervals and aggregation job volume such that the amount of noise is low.

> It would be better to provide ability to query for completed jobs. This enables leader to process-and-fetch the completed jobs in a loop.

That's an interesting idea. Can you elaborate on the structure of such an endpoint?

Reviewer 3: I think one challenge for "give me any complete job" is deciding when to have the Helper _stop_ returning a given job to this request type.

If the Helper returns a given job only once, an inopportune network issue, Leader crash, or other transient issue could lead to the job being "lost". (specifically, the condition for a job being "lost" would be the Helper considering the request successful, but the Leader not processing the response successfully)

If the Helper hands out the job until e.g. the Leader sends the next `AggregationJobContinueReq`, a Leader which sends multiple requests for completed jobs concurrently could see the same response multiple times. (Also, if we have reached the last aggregation step, there is no next `AggregationJobContinueReq` to be sent.)

Author: Something to note is that async init/continue raises interesting possibilities as to how many reports are contained within an `AggregationJobInitReq`.

Right now, the leader/helper have to limit the number of reports based on how long HTTP timeouts are configured helper-side.

With async, the leader can send arbitrarily large aggregation jobs without concerns about the HTTP gateway timeout. If taking advantage of that, there are fewer aggregation jobs that require bookkeeping and polling.

Author: Resolving this subthread: I feel strongly that per-job polling is sufficient, and that implementations can control the volume of GET polls by tweaking the aggregation job size and polling interval.

Feel free to unresolve if there are dissents.

aggregation job by sending requests to
`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` until
receiving HTTP status 200 OK with a body containing an `AggregationJobResp`.

> The response MAY include a Retry-After header field to suggest a polling

Reviewer: A bit more clear
```suggestion
The Helper's response MAY include a Retry-After header field to suggest a polling
```

Finally, the Leader sends a PUT request to
`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`. The payload
is set to `AggregationJobInitReq` and the media type is set to

> "application/dap-aggregation-job-init-req". After receiving HTTP status 202
> Created, the Leader proceeds to poll GET requests to the aggregation job URI

Reviewer 4: ```suggestion
"application/dap-aggregation-job-init-req". After receiving HTTP status 201
Created, the Leader proceeds to poll GET requests to the aggregation job URI
```

Reviewer 4: Further, the use of the verb "poll" is awkward: what's being polled is the aggregation job, not the GET request. The GET request is the representation of polling an aggregation job. Something like "the Leader polls the aggregation job by sending GET requests to the aggregation job URI" is better.

until receiving HTTP status 200 OK with a body containing an

> `AggregationJobResp`.

Reviewer 4: This new text should be moved a few lines below, to where it says "The Helper's response will be...".

First, it's awkward to discuss response handling and then have a sentence discussing Leader->Helper authentication, which is relevant to the request.

Second, the paragraph beginning "The Helper's response will be an `AggregationJobResp` message..." should be updated, because it's not clear whether that's talking about the PUT or a subsequent GET. 

Second,

* `prepare_inits`: the sequence of `PrepareInit` messages constructed in the
  previous step.

Finally, the Leader sends a request to

> `PUT {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`. The

Reviewer 4: nit: I don't love this phrasing. `PUT` is the verb in the request sent to the path, not the recipient of the request.

Author: I mostly took this to be consistent with the existing phrasing of https://www.ietf.org/archive/id/draft-ietf-ppm-dap-09.html#section-4.6.1-14, but now perusing the draft it looks like that was in fact an outlier usage of `PUT /blah`. I will rephrase as `sends a PUT request to "/blah"`.

Reviewer 4: IMO consistency is king here, but also it's not worth sweating this too hard, because I suspect that things like request paths might get revisited as DAP makes its way to RFC (specifically I suspect we'll get forced to do an ACME style directory).

payload is set to `AggregationJobInitReq` and the media type is set to
"application/dap-aggregation-job-init-req".

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

After receiving HTTP status 201 Created, the Leader proceeds to poll the
aggregation job by sending requests to

> `GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` until

Reviewer: Let's add the step parameter here as well ?
```suggestion
`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0` until
```

until receiving HTTP status 200 OK with a body containing an
`AggregationJobResp`. The Helper's response SHOULD include a Retry-After header

> field to suggest a polling interval to the Leader.

Reviewer 4: This last sentence suggests that the eventual 200 OK `AggregationJobResp` response should come with a `Retry-After` header. I think we should reorder the sentences.

* `prepare_inits`: the sequence of `PrepareInit` messages constructed in the
  previous step.

Finally, the Leader sends an HTTP PUT request to
`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` with a media
type of "application/dap-aggregation-job-init-req" and a body containing the
`AggregationJobInitReq`.

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

> The Helper responds with HTTP status 201 Created. If the response has a media

Reviewer 5: Shouldn't this response include a `Location` field indicating what was created?

Reviewer: I assume you mean https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Location?

Great question. We actually specify the path for the subsequent GET in this PR, so it might not be strictly necessary?

One potential oddness here: If the helper responds synchronously, i.e., the body of the response contains the `AggregationJobResp`, so there would be no subsequent GET, and therefore no Location to return. Perhaps the leader should respond with 200 OK in this case?

Author: I think no action is required here. Per https://www.rfc-editor.org/rfc/rfc9110#section-15.3.2-1

> The primary resource created by the request is identified by either a [Location](https://www.rfc-editor.org/rfc/rfc9110#field.location) header field in the response or, if no [Location](https://www.rfc-editor.org/rfc/rfc9110#field.location) header field is received, by the target URI.

since the PUT and GET URIs are the same, we shouldn't need a `Location` header. (please correct me if I'm misreading)

Author: Ah, my mistake, the PUT/POST and GET URIs are not actually the same since we don't require the step field in the PUT/POST request--in which case yes we should have a Location header.

type of "application/dap-aggregation-job-resp" with a body containing an
`AggregationJobResp`, the Leader proceeds onward (see
{{aggregation-helper-init}}). Otherwise, the Leader polls the aggregation job
by sending HTTP GET requests to

> `{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0`. If the

Reviewer 5: Where did `?step=0` come from?

Reviewer: "step" refers to the step of the aggregation job:

* step 0 is the initial `AggregationJobInitReq`
* step 1 is the first `AggregationJobContinueReq`
* step 2 is the second `AggregationJobContinueReq`
* and so on.

Note that, for Prio3, aggregation is complete after step 0.

We decided to add the `step` parameter to the GET request so that the helper knows how to respond if the leader somehow ends up in a weird state.

job is not ready, the Helper responds to the GET request with HTTP status 202

> Accepted with no media type and body. The Helper's response SHOULD include a

Reviewer 5: 202 is not the right status code to return from polling.  If the goal is to have a resource that indicates the status of the job, it should always return 200 with a body that indicates what the status is.

Reviewer: Got it, thanks for catching this.

When the step is ready, the Helper returns 200 OK with an `AggregationJobResp`, so we would need to disambiguate this "ready" response from a "not ready" response. Could we use the content types for this? Something like:

* "ready": content = [application/dap-aggregation-job-resp](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-8.1.4); body = [`AggregationJobResp`](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.7.1.2-19)
* "not ready": content = application/dap-aggregation-job-pending (or something); body = empty

Or perhaps it would be more correct to change `AggregationJobResp` to convey whether the step is ready?

Author: > 202 is not the right status code to return from polling. If the goal is to have a resource that indicates the status of the job, it should always return 200 with a body that indicates what the status is.

That tracks with prior art https://www.rfc-editor.org/rfc/rfc8555#section-7.5.1. (How embarrassing! I should have consulted ACME far earlier :wink:).

I have added a `status` field to the `AggregationJobResp`. I think this brings nice text simplifications, because we express the mutual exclusion between a processing job and the presence of prep_resps through the struct definition.

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

The Helper responds with HTTP status 201 Created. If the response has a body
containing an `AggregationJobResp`, the Leader proceeds onward (see
{{aggregation-helper-init}}). Otherwise, if the response has an empty body, the

> Leader polls the aggregation job by sending HTTP GET requests to

Reviewer: Instead of using empy/non-empty to distinguish the response, the Leader should use the media type. That is, if the response contains the media type header for AggregationJobResp (I don't remember what this is), then interpret the body of the response as an AggregationJobResp.

* `prepare_inits`: the sequence of `PrepareInit` messages constructed in the
  previous step.

Finally, the Leader sends an HTTP PUT request to
`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` with a media
type of "application/dap-aggregation-job-init-req" and a body containing the
`AggregationJobInitReq`.

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

The Helper responds with HTTP status 201 Created with a body containing an
`AggregationJobResp` (see {{aggregation-helper-init}}). If the `status` field
is `finished`, the Leader proceeds onward. Otherwise, if the `status` field is
`processing`, the Leader polls the aggregation job by sending GET requests to

> the URI indicated in the Location header field, until the `status` is

Reviewer: Why doesn't this work just like [polling the collection job URI](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.8.1-15)?

Reviewer: No need to change it, but if you think it's worth adjusting collection to match, then let's file an issue to track.

Author: Collection job polling actually runs awry of @martinthomson's feedback about 202 being the wrong code, needing Location header, etc. https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/564#discussion_r1701106596.

So I think we should adjust collection to match.

The Helper begins an aggregation job when it receives an `AggregationJobInitReq`
message from the Leader. For each `PrepareInit` conveyed by this message, the
Helper attempts to asynchronously initialize VDAF preparation (see
{{Section 5.1 of !VDAF}}) just as the Leader does. If successful, it includes
the result in its response that the Leader will use to continue preparing the
report.

Upon receipt of an `AggregationJobInitReq`, the Helper checks if it recognizes
the task ID. If not, then it MUST abort with error `unrecognizedTask`.

Next, the Helper checks that the report IDs in
`AggregationJobInitReq.prepare_inits` are all distinct. If two preparation
initialization messages have the same report ID, then the Helper MUST abort with
error `invalidMessage`.

If the Helper finds the `AggregationJobInitReq` to be valid, it immediately

> responds with HTTP status 201 Created.

Reviewer: Should this be 202?

Reviewer 4: No, 201 Created is what we use in response to `PUT /tasks/{task-id}/collection_jobs/{collection-job-id}`, so it's appropriate here. 202 Accepted is what we use in response to `GET /tasks/{task-id}/collection_jobs/{collection-job-id}` if the job is not done yet.

The Helper is now ready to process each report share into an outbound prepare
step to return to the Leader. Asynchronously from any further requests from the

> Leader, the following structures are computed:

Reviewer: This transition is a little awkward I think. I like the first paragraph ("it immediately responds with ..."). Perhaps tack on something like: "Later on, the Leader will make send an HTTP GET to <blah blah blah> to get the response. The Helper constructs the response as follows."

Author: I rewrote this to match what I did for continuation
```
If the Helper finds the `AggregationJobInitReq` to be valid, it immediately
responds with HTTP status 201 Created. The Helper is now ready to process each
report share into an outbound prepare step, asynchronously from any further
requests from the leader.

The following structures are computed:
```

LMK if that's also awkward, then I can fix both :D

Reviewer: Much better!

initialization messages have the same report ID, then the Helper MUST abort with
error `invalidMessage`.

If the Helper finds the `AggregationJobInitReq` to be valid, it MAY immediately

> respond with HTTP status 201 Created with no body, and perform the remaining

Reviewer: Mention media type here as well.

Author: I think it may be better if I strike `with no body`, as elsewhere in collection it's inferred that if we only say "responds with STATUS" there is no media type and body, as compared to explicitly saying "responds with STATUS with a body of SomeResp and media type foo/bar"

where `prepare_resps` are the outbound prep steps computed in the previous step.
The order MUST match `AggregationJobInitReq.prepare_inits`.

After receiving the response to its `AggregationJobInitReq`, the Leader makes

> an HTTP GET request to the aggregation job URI to check on the status of the

Reviewer: is "aggregation job URI" well-defined? In any case, I think it would help the reader to either spell it out here or add a reference to the definition.

Author: It's not concretely defined--I copied the vernacular from the collection flow. I think I will expand it out instead of using the shorthand (and do the same for the other references).

Author: Done. I've replaced mention of `aggregation job URI` and `collection job URI` with the explicit paths.

aggregation job. If the aggregation job is ready, the Helper responds with
HTTP status 200 OK and a body consisting of an `AggregationJobResp`, with media
type "application/dap-aggregation-job-resp". If the aggregation job is not
finished yet, the Helper responds with HTTP status 202 Accepted. The response
MAY include a Retry-After header field to suggest a polling interval to the
Leader.

Changing an aggregation job's parameters is illegal, so further requests to
`PUT /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the same
`aggregation-job-id` but with a different `AggregationJobInitReq` in the body

> MUST fail with an HTTP client error status code.

Reviewer 4: @cjpatton I believe this paragraph answers your question "How does the Helper multiple PUTs for the same aggregation job?"

PUT multiple times is legal for idempotence, but mutation is illegal.

where `prepare_resps` are the outbound prep steps computed in the previous step.
The order MUST match `AggregationJobInitReq.prepare_inits`.

After receiving the response to its `AggregationJobInitReq`, the Leader makes
requests to
`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` to check
on the status of the aggregation job. If the aggregation job is ready, the
Helper responds with HTTP status 200 OK and a body consisting of an
`AggregationJobResp`, with media type "application/dap-aggregation-job-resp".
If the aggregation job is not finished yet, the Helper responds with HTTP

> status 202 Accepted. The response MAY include a Retry-After header field to

Reviewer 2: It looks like this requires constant polling from the leader end to check status of each job. Should it implement long polling and respond when there is response to return?

Reviewer 2: If we are not doing long polling for any reason, the response should probably 200 with appropriate status response instead of 202, since we are not accepting a new request for processing.

Reviewer: What do you mean by "long polling"?

Author: Hmm, I see long polling as the state of the spec before this change, more or less. That is, leaders hold the HTTP connection open while the helper does processing. The helper returns a response when it's ready.

> If we are not doing long polling for any reason, the response should probably 200 with appropriate status response instead of 202, since we are not accepting a new request for processing.

202 signals to the leader that polling needs to continue. We could give 200 with an empty body, but I think 202 is most appropriate, [given my reading of MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/202). (FWIW I copied the semantics from collection, so if we argue against using 202 we would need to fix collection too).

Reviewer 2: Earlier model is, you wait when you init a job. Now, we submit the job and then wait on a different thread. But leader has to wait for several jobs to complete in this model, right? So, it is not clear to me how leader keep polling for all jobs better than long polling.

@cjpatton - http long polling - in this context, it translates to helper blocks the GET call until the job processing is done (with an optional timeout).

Author: The goal is to enable the helper to move expensive computation (checking uniqueness indices, heavyweight VDAF math, etc.) out of the request hot-path. 

It is plausible to make no changes to the spec, which IIUC is equivalent to using a long-polling strategy. Implementations could still decouple expensive operations by leaving the connection open and internally polling a pool of compute workers.

But I (if I'm operating as a helper) am still left with the disadvantages of having to deal with long-running connections. It's tricky to load balance them, e.g. what do we do with existing connections when we need to scale up? Whereas for stateless GET requests with no body, load balancing them is trivial.

If we really want to avoid polling, we may want to look into the more robust 2-way protocols, i.e. websockets.

Reviewer 2: To summarize my proposed changes - 
1. aggregate init request can return either 200 with proper response body (sync) or 202 without body (async). This makes it backward compatible and supports sync model.
2. When leader is ready to receive the init job response, it makes a GET <jobid> http request. Helper can
  * respond with 202 immediately if the job init is not done on helper (frequent polling model)
  * respond with 200 immediately with init job response (if job is completed)
  * Block the caller and respond with 200 with init job response when job is finished (long polling).

Reviewer 3: I agree that the old semantics were effectively equivalent to long-polling.

I also think that the semantics in this PR are compatible with a long-polling implementation: specifically, on receipt of a poll request (for an incomplete job) the Helper would not respond at all until the job is complete (resulting in a 200 response) or some timeout occurs (resulting in a 202 response).

But now that we're thinking about this, I do wonder if this could be a source of friction between different implementations: the deployment considerations for immediate-polling are different than those for long-polling, so an implementation optimized for immediate-polling may not play well with an implementation optimized for long-polling.

Reviewer 2: I hope the above clarifies some of the confusion. We are still doing it this process in two steps, but the later step can be optimized as mentioned above without compromising on proposed async model.

Reviewer 3: What textual change are you suggesting? My interpretation of the text in this PR is that an implementation is allowed to implement either frequent-polling or long-polling. There are some operational considerations which are not specified (e.g. if long-polling, how long does the Helper wait before giving up & returning 202?) but I think they can be left implementation-dependent.

N.B. as it stands, the initial request must always return a 201 Created -- i.e. the old "sync" model has been replaced by the new "async" model. The prior discussion around this is in #557.

Author: > My interpretation of the text in this PR is that an implementation is allowed to implement either frequent-polling or long-polling.

True, with caveats. The GET request that follow PUT/POST requests could long poll. But note that this PR changeset would have inserted a new request that didn't exist before, i.e. long polling on the PUT/POST requests is no longer possible.

But as @suman-ganta states, we can make it so that:
> aggregate init request can return either 200 with proper response body (sync) or 202 without body (async). This makes it backward compatible and supports sync model.

In effect we're raising the question again we thought about in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/557#issuecomment-2166938611. I do see value in letting the polling behavior (long or short) be decided amongst implementations, but it does seem to come at the cost of spec complexity.

Reviewer 3: I think we should be careful not to conflate the discussion of:

* Sync vs async aggregation job handling (and whether to support both of these, or only one), first opened in #557 and continued in this PR.
* Frequent-polling vs long-polling, first discussed in this PR.

FWIW/IMO, I am concerned about the need for an extra request per aggregation job implied by the async model, but not enough to accept the complexity of specifying & implementing both sync & async behaviors -- I think we should specify only one of sync & async. I am still thinking through whether it is wise to allow both frequent-polling & long-polling behaviors.

Reviewer: I think the PUT should always return 202. I like @suman-ganta's suggestion for the GET. I understand it to be:

> If the response is ready, then return 200. Otherwise, if the response is not ready, the Helper MAY return 202 immediately, or it MAY wait until the response is ready and respond with 200.

Reviewer: A textual change may make this clearer.

Author: > I think the PUT should always return 202.

Clarify? PUT should always return 201 Created?

> I like @suman-ganta's suggestion for the GET.

I'm fine with this change as well.

> A textual change may make this clearer.

Agreed, I think the behavior is permissible with the text as-is, but it's not obvious.

> field to suggest a polling interval to the Leader.

Author: For consistency, I've also called this out in collection as well.

Changing an aggregation job's parameters is illegal, so further requests to
`PUT /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the same
`aggregation-job-id` but with a different `AggregationJobInitReq` in the body

> MUST fail with an HTTP client error status code.

Reviewer 6: This is very hard to implement for the helper. It requires possibly storing every body indefinitely. I think we can take advantage of the fact that aggregation is now async to make this better.

I propose that subsequent `PUT` requests to this enpoint carrying the same aggregation-job-id should be responded with the same response that previous requests were responded with regardless of , this means we have to, at most, store an (endpoint, response code) tuple. And we still maintain the idenpotency guarantees that we want, in case the leader retries the request

Reviewer 3: I disagree here: in practice, the Helper only needs to store a hash of the previous request to perform request-equivalence checking. So the storage overhead isn't to store all request bodies indefinitely, it's to store the hash of the last request body until the next request is successfully processed.

I do think that checking that the request matches is very useful for "request/response coherency" reasons: if the Helper doesn't check the request body, we could end up in a situation where the Leader (due to a bug) repeats a request for some aggregation step with a different request, and the Helper would respond with the response body for the original request. This could be a very challenging situation to debug.

Author: > It requires possibly storing every body indefinitely.

Can you clarify the drawbacks to this? In Janus, [we accomplish this by storing a SHA-256 of the request body](https://github.com/divviup/janus/blob/main/db/00000000000001_initial_schema.up.sql#L225) which is fairly cheap in terms of storage, and I don't think has caused us problems (yet).

If we have to store `(endpoint, response code)` anyway, then we may as well go all the way and give more "truthful" idempotency (i.e. retain this clause), assuming that storing a request hash is not burdensome.

Reviewer 6: You're right, this can be implemented with a hash. My follow up question is, what happens if the leader does repeat the request and keeps the same body? What should the helper do then? If this is specified I'm missing where

Author: I'm not seeing concrete mention of that scenario either. We've been treating repeated requests as idempotent, i.e. we return 201 Created.

We probably need language in this paragraph along the lines of "If the `AggregationJobInitReq` is identical to the original request, the Leader MAY respond with 201 Created". Thoughts from others?

Reviewer 4: My view is that the discussion of PUT in [RFC 9110](https://www.rfc-editor.org/rfc/rfc9110#section-9.3.4) makes it clear that such request are idempotent, but there's no downside to making that explicit here.

Reviewer: @inahga let's mention this scenario explicitly.

where `prepare_resps` are the outbound prep steps computed in the previous step.
The order MUST match `AggregationJobInitReq.prepare_inits`.

After receiving the response to its `AggregationJobInitReq`, the Leader makes
requests to
`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` to check
on the status of the aggregation job. If the aggregation job is not finished

> yet, the Helper MAY wait for the job to complete. Once the aggregation job is

Reviewer: ```suggestion
yet, the Helper MAY wait to respond until the job is complete. Once the aggregation job is
```

> ready, or if it is already ready, the Helper responds with HTTP status 200 OK

Reviewer 7: Once a job is ready for collection, do _all_ polls have to return `200 OK` consistently?

I ask because the usage of eventually consistent mechanisms could lead to polls even after one which returned 200 OK sometimes indicating that the job isn't ready.

Author: I don't think a poll has to return 200 OK consistently for the protocol to work. Extra polls are harmless--the leader will simply try again. Or perhaps, whether it's harmless depends on the consistency guarantees between leader and helper, which I don't think DAP is prescriptive of right now.

I also don't think that behavior runs awry of the text. Perhaps we could disambiguate by saying `the Helper SHOULD respond with...` but that's a bit awkward since it implies a compliant aggregator could never return 200 OK... (To be clear, I think the behavior you describe should be allowable).

Reviewer 7: I didn't think that behavior was proscribed by the text either, but wanted to verify.

It's probably worth calling this out as a possibility in the text as a heads up to leader node implementors, though I'm not sure why a leader would try and poll again after seeing a completion.

Reviewer: +1 to calling this out. No normative language required, just something like "At least one GET will get the AggregationJobResp" in return.

Author: I've been staring at this for a while and am having difficulty succinctly writing it out, without it sounding confusing.

The closest I've gotten so far is to add a sentence `For each aggregation job init/continue, the Helper will respond at least once with HTTP status 200 OK.`, but I'm not sure that actually covers the scenario we've painted.

Also the more I stare at it the less I'm onboard with taking it. I don't think a Leader will needs to take any different action or encode any different logic to cover this case. The only situation I can think of where this case comes up is where the 200 response is lost in transit, the leader queries again and receives 202. In which case, the solution is to proceed normally until 200 is actually received.

where `prepare_resps` are the outbound prep steps computed in the previous step.
The order MUST match `AggregationJobInitReq.prepare_inits`.

If the Helper is asynchronously processing the aggregation job, the Leader makes
HTTP GET requests to
`{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0` to check
on the status of the aggregation job. If the aggregation job is not finished yet,
the Helper MAY wait to respond until the job is complete. Once the aggregation
job is ready, or if it is already ready, the Helper responds to the GET request
with HTTP status 200 OK and a body consisting of the `AggregationJobResp`, with
media type "application/dap-aggregation-job-resp". If the aggregation job is not
finished yet, and the Helper chooses not to wait for the job to finish, the

> Helper responds with HTTP status 202 Accepted. The Helper's response SHOULD

Author: I'm cool with it. I will save changing collection's wording for another PR to keep this one from continuing to sprawl. (In fact, I should probably back out all changes to collection and save them for a later editorial PR).

include a Retry-After header field to suggest a polling interval to the Leader.

Otherwise, if the Helper is synchronously processing the aggregation job, the
Helper responds to the PUT request with HTTP status 201 Created and a body
consisting of the `AggregationJobResp`, with media type
"application/dap-aggregation-job-resp".

Changing an aggregation job's parameters is illegal, so further HTTP PUT
requests to `/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the
same `aggregation-job-id` but with a different `AggregationJobInitReq` in the
body MUST fail with an HTTP client error status code. For further requests with
the same `AggregationJobInitReq` in the body, the Helper MUST respond with HTTP

> status 201 Created.

Author: I think idempotency is a required property of a [PUT request](https://www.rfc-editor.org/rfc/rfc9110#section-9.3.4-2), rather than related to skew recovery.

We could make it a SHOULD. The helper's behavior then should respond with an error code though, I don't think it's valid for the helper to accept more than one valid `AggregationJobInitReq`.

> status 201 Created.

Author: I've turned it into a SHOULD. I suppose for many-step VDAFs, the helper could want to discard its response to `AggregationJobInitReq` as later steps take place, i.e. `n-2, n-1, n`, `n-2` may be zero and can be discarded/overwritten while still supporting skew recovery. In which case the helper would respond with something like 410 Gone.

> status 201 Created.

Author: > Not for this PR, but perhaps that section should be generalized, since the initialization request is effectively "step 0".

I'm hoping in the next PR that section can be stricken entirely, now that we have step encoded in the URI. Although I haven't thought about it in the face of supporting both sync/async.

} AggregationJobStatus;

struct {
  AggregationJobStatus status;
  select (AggregationJobResp.status) {
    case processing: Empty;
    case finished:   PrepareResp prepare_resps<0..2^32-1>;
  };
} AggregationJobResp;
~~~

where `prepare_resps` are the outbound prep steps computed in the previous step.
The order MUST match `AggregationJobInitReq.prepare_inits`.

The Helper responds to the Leader with HTTP status 201 Created, a body
consisting of the `AggregationJobResp`, and the media type
"application/dap-aggregation-job-resp".

Depending on the task parameters, processing an aggregation job may take some
time, so the Helper MAY defer computation to a background process. It does so
by responding with the field `status` set to `processing` and a Location header
field set to the relative reference
`/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step=0`. The Leader
then polls the Helper by making HTTP GET requests to the aforementioned
Location. The Helper responds to GET requests with HTTP status 200 and the
`status` field reflecting the current state of the job. When the aggregation
job is `processing`, the response SHOULD include a Retry-After header field to

> suggest a polling interval to the Leader.

Reviewer: Consider spelling out the status code if a GET is made before the PUT. This might happen for example if the Leader crashes during the polling loop.

Reviewer: Here and below (continuation).

Author: Is this scenario assuming that a Leader moves onto polling without check-pointing its progress? That is, it does
```rust
let mut aggregation_job_resp = post(aggregation_job);
while !aggregation_job_resp.status.is_done() {
    // n.b. no checkpointing to datastore is done
    aggregation_job_resp = get(aggregation_job_resp.polling_location);
}
```
So if it crashes during this function, it'll restart from the `post()` call?

I think this is already specified more or less by:
```
For further requests with
the same `AggregationJobInitReq` in the body, the Helper SHOULD respond as it
did for the original `AggregationJobInitReq`, or otherwise fail with an HTTP
client error status code.
```

Since it'll be repeating the AggreationJobInitReq.

(Let me know if I'm misunderstanding the scenario).

Reviewer: I was thinking more of a misbehaving Leader trying to GET an aggregation job step before kicking it off with a PUT/GET. Thinking about it more, this is just a plain old 404. I don't think it's worth documenting.

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

After receiving HTTP status 202 Accepted, the Leader proceeds to poll the
aggregation job by sending requests to

> `GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` until

Reviewer: The same URI is used for each prep step. Do we foresee this being an issue? Perhaps this is actually desirable from an idempotency standpoint?

cc/ @tgeoghegan, @branlwyd

Reviewer 4: It was already the case that the same URI was used regardless of which step we were on. Asynchronous handling of aggregation jobs shouldn't change that.

Reviewer 2: Since this is async interaction, leader can send two POST requests to this URI, one with valid and another with invalid body. When we query for status with GET, it would be ambiguous to interpret the response. Two options to resolve this -
* Maintain a requestID so a GET can be pointed to a specific requestID
* Somehow early reject 2nd request instead of always returning 202

Reviewer: Invalid in what sense? If the message is malformed or something, then the Helper would abort the request. I think it's the Helper's job to ensure that one and only one request is accepted.

Reviewer 2: When two requests are accepted with 202 on the same URL, there is no way to know the response fetched later belongs to request1 or request2. Hope that clarifies.

Reviewer: When would this happen? Are you thinking of something like a race condition, where the Leader erroneously sends two POSTs at the same time for the same aggregation job?

Reviewer 2: It could happen for various reasons -
1. Leader did not receive response due to network error and retries.
2. Leader incorrectly tracks the step count and sends continue request with invalid step count.
3. Two leader instances sends the same request
4. Other malicious reasons

I think without request to response correlation, it adds stateful processing responsibility on leader and still does not give a way to track this correlation in a consistent way.

Reviewer 3: Given that the relevant VDAF functions are deterministic, when both aggregators are honestly following the protocol, there is exactly one correct next request for a given aggregation job[1].

Therefore, I suspect the semantics we want are that the _first_ request received for a given aggregation request determines the correct request; if the Leader sends multiple, non-identical requests for the same aggregation job, something has gone wrong with the Leader & failure is OK. That is, we want to detect repeated non-identical requests, but the proper way to respond to them is with failure rather than tracking both requests.

[1] This is true semantically but may not be true up to encoding; though DAP messages are deterministically-encoded if the underlying VDAF messages are, and to my knowledge all current VDAFs use messages which are deterministically-encoded.

Reviewer: VDAF preparation is deterministic. Given the current _strict_ ordering of reports in the job, there should be exactly one valid next message.

Reviewer: Well, one thing that can change is that the Helper's decryption key gets rotated out, causing the report to be rejected.

Reviewer 2: I guess I don't get what response Helper gives in these cases. Response to last request or response to first request?

If it is giving response to first request to all - then an incorrect submission (request2) gets a valid response.
If it is giving response to last request to all - then a valid submission receives error response without knowing why.

Besides this, My understanding is, leader does not have to be good citizen and can learn things and hence Helper implements
* Prevention of replay attacks
* Prevention of same share belonging to several batches etc.
Is that right understanding and still holds good here?

Reviewer: Yes, both the Leader and Helper are responsible for preventing replays, including across batches.

Reviewer 2: Okay. I still need to understand this part -

> 
> I guess I don't get what response Helper gives in these cases. Response to last request or response to first request?
> 
> If it is giving response to first request to all - then an incorrect submission (request2) gets a valid response.
> If it is giving response to last request to all - then a valid submission receives error response without knowing why.

Reviewer: >     1. Leader did not receive response due to network error and retries.

If the body of each request is the same, then so should be the Helper's respond. Otherwise, the Leader is in error, and it's up to the Helper to decide which order they're processed in.

>     2. Leader incorrectly tracks the step count and sends continue request with invalid step count.

The Leader is in error. The Helper will abort, but it [might be possible to retry](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-4.5.2.3).

>     3. Two leader instances sends the same request

Same as for 1 I think.

>     4. Other malicious reasons

Even if the Leader is malicious, then it's still up to the Helper to enforce invariants like anti-replay. However there are probably any number of ways the Leader can break correctness.

Reviewer 2: If two mutating requests results in two different responses, then it is not deterministic to let Helper decide what to do with them. I think it is deterministic if the helper respond with the unique URL to track the POST request status. I think this behavior should not depend on what Leader's behavior is.

Note that, this would not be a problem if the response is synchronous since it comes with request-response correlation built-in.

Reviewer: Let's maybe take a step back for a second: what property do you want DAP to have that this PR doesn't have? It sounds like you want some way for the Leader to recover if it makes a mistake.

It sounds like you're suggesting the following mechanism: the Leader chooses a unique identifier for each request (probably at random?), which the Helper is supposed to echo back in its response. Is this what you're suggesting?

If so, what errors does this help the Leader recover from? What errors does it not help it recover from?

Author: > When two requests are accepted with 202 on the same URL, there is no way to know the response fetched later belongs to request1 or request2.

Concretely, in Janus, each insertion of aggregation job is transactional since we use a relational database. The operations are serialized--a second request will reject with a database conflict error. That is, the only scenario where we would return 202 for a repeated request is if the second request is exactly identical to the first.

However, I don't think we have any requirement for transactionality, either in this PR or generally in DAP. An aggregator that lacks transactionality _could_ accept multiple PUTs, and indeed it would be implementation-defined as to which request the subsequent GET request is responding to.

So I think @suman-ganta is correct. Perhaps we should add clarification that the helper should only accept one PUT request?

Author: Eh. Reading through it again as I try to write it:
```
Changing an aggregation job's parameters is illegal, so further requests to
`PUT /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for the same
`aggregation-job-id` but with a different `AggregationJobInitReq` in the body
MUST fail with an HTTP client error status code.
```
I feel we are adequately signaling that only one PUT request should be accepted. i.e. a helper that is accepting multiple different PUT requests is not following the spec. Similarly, I think the continue section with step recovery strongly implies the need for transactionality/consistency.

So I think my point earlier:
> However, I don't think we have any requirement for transactionality, either in this PR or generally in DAP.

is not valid. Sorry, I should have given a close read before commenting.

So @suman-ganta, I think the questions in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/564#discussion_r1679566081 still apply.

---

> Also @suman-ganta and @kristineguo point out offline that, if a POST comes in for step 2 while we're processing step 1, then we should abort the POST request.

I think stepMismatch covers this? If the helper receives a request for step 2 while step 1 is still in flight, the aggregation job will still be on step 1, thus being errored out per this paragraph.
```
Next, the Helper checks if the continuation step indicated by the request is
correct. (For the first `AggregationJobContinueReq` the value should be `1`;
for the second the value should be `2`; and so on.) If the Leader is one step
behind (e.g., the Leader has resent the previous HTTP request), then the Helper
MAY attempt to recover by sending HTTP 202 status Accepted and discontinuing
any further work. In this case it SHOULD verify that the contents of the
`AggregationJobContinueReq` are identical to the previous message (see
{{aggregation-step-skew-recovery}}). Otherwise, if the step is incorrect, the
Helper MUST abort with error `stepMismatch`.
```

Reviewer 8: I don't know how often this scenario would actually come up in reality, but imagine a scenario where the Leader sends out two synchronous requests when recovering:
- `POST /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for step 1 (already computed)
  - Helper may return `202 Accepted` here if the request matches the previous message.
- `POST /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` for step 2 (new)

Does the Helper return `202 Accepted` for that second POST request? If the Helper does, what should the Helper return when the Leader does `GET /tasks/{task-id}/aggregation_jobs/{aggregation-job-id}`? The response for step 1 or 2?

Reviewer 8: To me, the confusion can arise because we're using the same request path for multiple different requests (one for each step). We could potentially alleviate this by making the request path more explicit by adding the step into the requestPath, so that the Helper always knows which step the Leader is sending/requesting.

Author: Good scenario, I think it's plausible. Perhaps not under normal operations, but it is a series of operations that I think would be accepted. Let me think about it out loud:

I'm going to assume that aggregation jobs processing is perfectly transactional, and that these simultaneous requests don't leave the job in some indeterminate state. I believe these are valid assumptions, but stop me if they're not.

---
Consider the first scenario, where the order of requests is serialized to request 1 (i.e. your bullet point 1), then request 2.

Request 1: POST returns 202 Accepted, any subsequent GET requests _would_ return 200 with the proper body, since it was already computed.
Request 2: POST returns 202 Accepted, any subsequent GET requests would return 202 until the aggregation job is complete.

I don't think anything is surprising here.

---
Second scenario, the order of requests is flipped.

Request 2: POST returns 202 accepted, any subsequent GET requests would return 202 until the aggregation job is complete.
Request 1: Step recovery kicks in, because we're trying to POST for step 1. Uh oh. I think you're right that this situation is ambiguous. For step recovery to actually be possible, we'd need to abort continuation to step 2. Or the leader would need to wait for step 2 to be done, then try step recovery again.

---

> To me, the confusion can arise because we're using the same request path for multiple different requests (one for each step). We could potentially alleviate this by making the request path more explicit by adding the step into the request path, so that the Helper always explicitly knows which step the Leader is requesting a response for.

I had originally considered having different paths for different steps, but couldn't find a scenario where this was strictly necessary, but I think you've identified one.

Let me dwell on it for a bit, anyone else have thoughts as well?

Reviewer 7: @inahga I don't think that we should be prescribing serializable transactions here - this is very limiting and operationally challenging for partitioned helpers.

With our helper, we've found that this aspect of the replay checks are the hardest part of the system to scale up - they eventually become a breaking point for the system unless you shard, and that has availability implications under draft 11 due to the synchronous nature of the protocol.

Ideally we'd make things idempotent, or defer when transactional things have to occur until after jobs have been accepted for processing.

Author: Fair points. It sounds like then simultaneous POSTs with different steps would be exceptionally painful for you, because it'll become indeterminate which request is being responded to for the subsequent GET. Does that sound right?

Reviewer 2: An alternative to step specific path is to return 202 with a trackable URL. This URL contains unique ID in it generated by the helper that leader can GET. It responds with status of that specific submission.

Another simpler way is to synchronously check if there is a pending `continue` submission to be processed for the job, if so, abort synchronously with an error.

Author: Those resolutions all seem plausible to me.

OK to summarize the thread direction: sharing the same URI for each step introduces ambiguity as to which request takes priority when requests for different steps come in simultaneously.

The solutions to this are:
1. Per step request routing, e.g. `{task-id}/aggregation_jobs/{aggregation-job-id}/{step-number}`.
1. Provide trackable URLs per PUT/POST, such that subsequent GETs can use it to get the proper response.
1. Add prerequisite that aggregators MUST abort if another init/continue is still in progress.
1. Do nothing--the scenario we painted above is admittedly somewhat pathological.

I favor solution one, because I think it'll simplify step recovery substantially, because the leader can straightforwardly GET step it has lost track of.

Does anyone else who hasn't chimed in in a while have thoughts, @branlwyd @cjpatton?

Reviewer 7: You've got it!

Reviewer: Thanks @inahga for the super succinct summary!

With the major caveat that I've not implemented the Leader: I'm inclined to take 4. If I understand correctly, this case occurs because something has gone seriously wrong with the Leader. Generally speaking, [we have stricter operational requirements for the Leader than Helper](https://ietf-wg-ppm.github.io/draft-ietf-ppm-dap/draft-ietf-ppm-dap.html#section-5.1.2-1). Thus I think we could reasonably "give up" and accept the data loss in this case.

That said, those that have implemented the Leader get a bigger vote share. I would nudge us towards 1 as it seems like the simplest way to get what we need. We might tweak it a bit to maintain the "resource API" semantics: `{task-id}/aggregation_jobs/{aggregation-job-id}?step={step-number}`

Reviewer: It also seems like this will be helpful for https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/569.

Reviewer 3: I like #3 or #1.

#3 feels the cleanest since, in normal operation, only a single aggregation step will be computed at once; it's OK to fail if multiple steps are attempted at once. The step would be considered "complete" (allowing another step to be computed) once the Helper considers its response to be retrieved at least once.

I'm not sure if #3 would step on the toes of some eventual-consistency scheme, however. If so, #1 seems the next cleanest.

Reviewer 6: I also agree with @cjpatton that for this scenario to happen the leader must be in some weird state, but I have also not implemented a leader. My thinking is that the leader would never even send a request for step 2 before having successfully polled the response for step 1 to completion, meaning the race condition could never occur. And as a helper, when I receive a request for step 2 after being done with step 1 I would just start aggregating step 2 and discard step 1 data. Meaning it would not be ambiguous for the helper which step the `GET` would be related to.

Now, despite this confusion on my part, I would still vote for a solution which includes the step number in the request (number 1), since it probably makes debugging/observability easier because it makes the intent of the leader obvious

Author: One curiosity about solution 1 is that it makes the step number in the `AggregationJobContinueReq` somewhat redundant. We may be tempted to take a wire breaking change to prune that field, as I don't think we would want to keep vestigial fields around.

---

> I'm not sure if 3 would step on the toes of some eventual-consistency scheme, however.

I had assumed so, hence my vote for 1. Each replica of an aggregator needs to have a perfectly consistent view of aggregation job status to be able to determine whether an aggregation job is in progress.

In some architectures, I believe this is really, really tricky. Naively I may implement async jobs as POST/PUTs pushing onto a message queue. Some dedicated message queueing software do not support introspecting the queue, e.g. [RabbitMQ](https://lists.rabbitmq.com/pipermail/rabbitmq-discuss/2011-August/014270.html). Some do, e.g. Sidekiq, but I believe that's because it's backed by a RDBMS.

So by imposing the requirement of 3, we have limited the implementation selection to software that can introspect the queue. Which, without having looked into it much, may only be well supported by relational datastores. _Even then_, depending on the sharding scheme being used, would a single request require introspecting the queues of all the shards?
---

I agree with all folks that the scenario only arises in the presence of a badly malfunctioning leader. But I think the fact that both requests can return 2xx responses as written in the spec is problematic, since only one request is obviously correct (i.e. the move to step 2), so I'm not favoring Do Nothing atm.

Reviewer 8: If you don't want to make a wire change, we could do:
`PUT {task-id}/aggregation_jobs/{aggregation-job-id}`
`GET {task-id}/aggregation_jobs/{aggregation-job-id}?step={step-number}`

The biggest issue is the `GET` has no way to explicitly identify the response it wants to retrieve. Leaving it up to the Helper to implicitly assume which step the Leader wants a response for leaves opportunity for (admittedly contrived) scenarios such as the one we've been discussing. But simply adding an explicit resource identifier of some sort, we avoid any potential for this confusion.

Hence why I'm partial to no.1, with step either as a query parameter or path component.

Reviewer: @inahga it sounds like 1 is the option with the most consensus. Let's take it for this PR.

Author: > If you don't want to make a wire change, we could do:

I meant the definition of `AggregationContinueResp`. It's defined as.
```
struct {
  uint16 step;
  PrepareContinue prepare_continues<1..2^32-1>;
} AggregationJobContinueReq;
```

We may be tempted to remove the `step` field, since if `step` is a field of the URI, it's no longer necessary. That would be a simplification, but a wire breaking change.

Author: > If you don't want to make a wire change, we could do:

Sorry, I have Monday brain, I see what you're getting at now. We could let `AggregationJobContinueReq` retain the `step` field, retain calling `POST` as-is, but only use the `step` query parameter in the GET request. That would not be a wire breaking change.

Author: I've sketched out how this looks in the latest commit, editorial changes still probably required. I went with striking the `step` field, mostly for consistency (don't need the same data in two different locations), and debug-ability (easier to see what the leader is thinking about through HTTP status logs).

I've also moved from POST to PUT for continuation, since we use separate URIs for each step. This lets us use PUT semantics, i.e. idempotency.

That idempotency leads me to strike the entire job recovery section. Now I believe recovery is implicitly possible. If the Leader loses the acknowledgement of the PUT request, it can simply retransmit. Likewise with losing the response to the GET requests. Hopefully that's not overeager, as that seems like a pretty big win for simplicity of the text.

I've replaced it with a small paragraph that allows the helper to discard information about old steps, so that it doesn't have to retain the entire step history of a (theoretical) VDAF with many steps, but allows for the flexibility of recovering the whole step history (a-la https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/569). I anticipate implementing this as `if step < job.step - 1 { return Err }`.

Reviewer: @inahga this is a big and somewhat contentious change, and for that reason, we need to make sure to implement the smallest change necessary. Please revert all changes except for adding the step parameter to the URL.

You are probably right about the simplifications we can make, but it will help everyone if we handle this in separate PRs. Feel free to just stick those in a draft PR for now.

Author: Makes sense, I have done so.

Reviewer: I'm closing this thread! @kristineguo or @suman-ganta please chime in if there's more to address.

receiving HTTP status 200 OK with a body containing an `AggregationJobResp`.
The response MAY include a Retry-After header field to suggest a polling
interval to the Leader.

> The response's `prepare_resps` must include exactly the same report IDs in the

Reviewer: Not part of this PR, but I think this should be be a proper MUST.
```suggestion
The response's `prepare_resps` MUST include exactly the same report IDs in the
```

same order as the Leader's `AggregationJobContinueReq`. Otherwise, the Leader
MUST abort the aggregation job.

> [[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
> 

Reviewer: Since you also deleted this above
```suggestion
```

The Leader MUST authenticate its requests to the Helper using a scheme that
meets the requirements in {{request-authentication}}.

After receiving HTTP status 202 Accepted, the Leader proceeds to poll the
aggregation job by sending requests to
`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step={step}`
where `step` is the `step` field of the `AggregationJobContinueReq`. It polls
until receiving HTTP status 200 OK with a body containing an
`AggregationJobResp`. The Helper's response SHOULD include a Retry-After header

> field to suggest a polling interval to the Leader.

Reviewer 4: As above, this suggests the 200 OK should contain `Retry-After`. This sentence should be attached to the text describing the 202 Accepted response.

Next, it checks if it recognizes the indicated aggregation job ID. If not, it
MUST abort with error `unrecognizedAggregationJob`.

Next, the Helper checks if the continuation step indicated by the request is
correct. (For the first `AggregationJobContinueReq` the value should be `1`;
for the second the value should be `2`; and so on.) If the Leader is one step
behind (e.g., the Leader has resent the previous HTTP request), then the Helper
MAY attempt to recover by sending HTTP 202 status Accepted. In this case it
SHOULD verify that the contents of the `AggregationJobContinueReq` are
identical to the previous message (see {{aggregation-step-skew-recovery}}).
Otherwise, if the step is incorrect, the Helper MUST abort with error
`stepMismatch`.

Next, the Helper checks whether the aggregation job is already in progress from
a prior continuation or initialization request. If so, it MUST abort with error

> `stepMismatch`.

Reviewer 4: Can you sketch out the scenario this is meant to catch? Per the previous paragraph, it should be OK for the leader to send the same `AggregationJobContinueReq` twice for idempotency reasons.

Author: Ah, you're right. I was thinking this would handle the case where the leader sends two different POST requests, but as you stated the idempotency check would catch this.

I will strike this step, and revert the reordering of steps (as it's not necessary).

If any of these checks fail, then the Helper MUST abort with error
`invalidMessage`. Additionally, if any prep step appears out of order relative

> to the previous request, then the Helper MUST abort with error

Reviewer 8: Why do we enforce the same ordering between requests for a given job? This seems unnecessarily strict when the Helper can just match based on the report_id.

Author: I think the discussion in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/217 is relevant.

`invalidMessage`. (Note that a report may be missing, in which case the Helper
should assume the Leader rejected it.)

Next, the Helper checks if the continuation step indicated by the request is
correct. (For the first `AggregationJobContinueReq` the value should be `1`;
for the second the value should be `2`; and so on.) If the Leader is one step
behind (e.g., the Leader has resent the previous HTTP request), then the Helper

> MAY attempt to recover by sending HTTP 202 status Accepted and discontinuing

Reviewer 8: ```suggestion
MAY attempt to recover by sending HTTP status 202 Accepted and discontinuing
```

> any further work. In this case it SHOULD verify that the contents of the

Reviewer 8: Can you clarify what you mean by discontinuing further work? Future continue steps should be allowed to continue processing asynchronously.

Author: I mean discontinuing work in the context of that single request, i.e. in the code it would be effectively a "return" statement. I will reword for clarity.

`AggregationJobContinueReq` are identical to the previous message (see
{{aggregation-step-skew-recovery}}). Otherwise, if the step is incorrect, the
Helper MUST abort with error `stepMismatch`.

If the Helper finds the `AggregationJobContinueReq` to be valid, it immediately
responds with HTTP status 202 Accepted. The Helper is now ready to continue
preparation for each report, asynchronously from any further requests from the

> Leader.

Reviewer 8: Same nit as above - this wording frames it as if the preparation the reports are asynchronous from each other, when it's really the jobs as the base unit of asynchronicity

(Note that a report may be missing, in which case the Helper should assume the
Leader rejected it.)

Next, the Helper checks if the continuation step indicated by the request is
correct. (For the first `AggregationJobContinueReq` the value should be `1`;
for the second the value should be `2`; and so on.) If the Leader is one step
behind (e.g., the Leader has resent the previous HTTP request), then the Helper

> MAY attempt to recover by sending HTTP 202 status Accepted. In this case it

Reviewer: What is the Helper trying to accomplish here? Does this response actually hep? If I understand correctly, the Helper's response would be the same had the Leader indicated the correct step.

Author: > if the Helper receives a repeated request it must treat it as if it needs to be handled asynchronously (since there is no way to communicate synchronous completion), though I suspect the implementation would effectively be to treat the job as already completed & return the results on the first attempt to poll.

That's more or less correct. My idea is that the repeated request is effectively a no-op, i.e. the helper-side aggregation job state does not need to change.

I've added verbiage
> If the Leader is one step behind (e.g., the Leader has resent the previous HTTP request), then the Helper MAY attempt to recover by sending HTTP 202 status Accepted and discontinuing any further work.

to cut the transition to the next paragraph. Let me know if more elaboration is necessary.

The Helper begins each step of continuation with a sequence of `state` objects,
which will be `Continued(prep_state)`, one for each report in the candidate set.

The Helper awaits an HTTP POST request to

> `{helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}` from the

Reviewer: Did we decide not to include the step in the path? E.g., `?step=<step>`?) Or did we decide to punt to a future PR?

Author: Since the POST body contains the step, a `?step=<step>` parameter is redundant. It's up to us whether we want to keep it in the body or as a query parameter, but I don't think we should do both.

IMO I don't think any particular one is better than the other.

If any of these checks fail, then the Helper MUST abort with error
`invalidMessage`. Additionally, if any prep step appears out of order relative
to the previous request, then the Helper MUST abort with error
`invalidMessage`. (Note that a report may be missing, in which case the Helper
should assume the Leader rejected it.)

Next, the Helper checks if the continuation step indicated by the request is
correct. (For the first `AggregationJobContinueReq` the value should be `1`;
for the second the value should be `2`; and so on.) If the Leader is one step
behind (e.g., the Leader has resent the previous HTTP request), then the Helper

> MAY attempt to recover by sending HTTP 202 status Accepted without performing

Reviewer 3: Similar to the other comment on skew-recovery/idempotency, this wording makes it sound like skew-recovery for continuation should always be done asynchronously (since synchronous recovery would respond with HTTP 200 OK). I don't think we should restrict the Helper in this way.

} PrepareResp;
~~~

After receiving the response to its `AggregationJobContinueReq`, the Leader
makes an HTTP GET request to the aggregation job URI to check on the status
of the aggregation job. If the continuation is not finished yet, the Helper
responds with HTTP status 202 Accepted. The response MAY include a Retry-After

> header field to suggest a polling interval  to the Leader.

Reviewer: I may have missed it, but this should apply to the initial agg request as well.

> header field to suggest a polling interval  to the Leader.

Reviewer 4: ```suggestion
header field to suggest a polling interval to the Leader.
```

If the aggregation job is ready, the Helper constructs an `AggregationJobResp`
message ({{aggregation-helper-init}}) with each prep step. The order of the

> prep steps MUST match the Leader's `AggregationJobContinueReq`. It responds

Reviewer 4: Interesting implication of this is that it means the helper must keep track of the order in which prep steps occurred in the request. Janus already does this, but perhaps there are implementations that relied upon having the leader request in memory.

Author: A point in favor of https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/217, but I doubt that discussion needs to be rehashed.

Reviewer: I don't think this requirement is any harder to meet as a result of taking this PR. Or am I missing something?

Reviewer 3: Hmm. In the Leader Continuation section, we have:

> The PrepareContinues MUST be in the same order as the previous aggregate request.

But in the Helper Continuation section, we have:

> Additionally, if any prep step appears out of order relative to the previous request, then the Helper MAY abort with error invalidMessage.

That is, we proscribe that the Leader MUST respect the original order of reports, but only that the Helper MAY respect the original order (and therefore MAY not). I'm not sure if there is a good reason for this mismatch -- it seems to me that this is Postel's Law ("be conservative about what you send, liberal in what you accept") thinking which is recognized as [causing reliability & security issues](https://en.wikipedia.org/wiki/Robustness_principle#Criticism). IMO, the Helper should require that reports be in the same order if the Leader does, i.e. the quoted MAY should be a MUST instead.

(That said, I do think the ordering requirement is perhaps ill-justified, but it's not important enough to rehash #217.)

Reviewer: I agree it's an oversight. It should be MUST.

} PrepareResp;
~~~

After receiving the response to its `AggregationJobContinueReq`, the Leader
makes requests to
`GET {helper}/tasks/{task-id}/aggregation_jobs/{aggregation-job-id}?step={step}`
to check on the status of the aggregation job. If the continuation is not
finished yet, the Helper MAY wait to respond until continuation is complete.
Once the continuation is ready, or if it is already ready, the Helper responds
with HTTP status 200 OK and a body consisting of an `AggregationJobResp`, with
media type "application/dap-aggregation-job-resp".

If the continuation is not finished yet, and the Helper chooses not to wait for
the job to finish, the Helper responds with HTTP status 202 Accepted. The
Helper's response SHOULD include a Retry-After header field to suggest a
polling interval to the Leader.

> If the aggregation job is ready, the Helper constructs an `AggregationJobResp`

Reviewer: ```suggestion
If the response is ready, the Helper constructs an `AggregationJobResp`
```

`AggregationJobResp` response to the Leader gets dropped due to something like a
transient network failure. The Leader could then resend the request to have the
Helper advance to step `n+1` and the Helper should be able to retransmit the
`AggregationJobResp` that was previously dropped. To make that kind of recovery

> possible, Aggregator implementations SHOULD checkpoint the most recent step's

Reviewer 8: Can we make this plural to indicate that Helpers can store anywhere from the most recent step to the all the steps in a job's history?

Author: Clarify? The Helper can store whatever it wants, but the minimum recommendation is to store whatever it needs to satisfy skew recovery, as far as the protocol is concerned.

Reviewer 8: The language implies that only the most recent step's state SHOULD be stored. I think we can generalize this language to imply that _all_ steps' state SHOULD/MAY be stored so promote aggregation step skew recovery.

Author: Or in other words, the Helper could retain the entire step history and satisfy the SHOULD, but I'm not sure that possibility needs to be explicitly called out in protocol text.

Reviewer 8: In general, the aggregation step skew recovery section seems to focus on just recovering when there's a difference of 1 between the Helper and Leader. I think it would be beneficial to change the wording to imply that aggregation step skew recovery can allow the Leader to re-send any number of previously computed steps. Hence, specifying the Helper can store all previous steps' state follows naturally from this.

Author: > The language implies that only the most recent step's state SHOULD be stored.

I don't think I agree with that reading, but we could say "Aggregator implementations SHOULD, at minimum, checkpoint...".

(n.b. this diff got roped into this PR because I noticed a typo in `AggregationJobResp`. We may consider taking this editorial into another PR).

Author: > I think it would be beneficial to change the wording to imply that aggregation step skew recovery can allow the Leader to re-send any number of previously computed steps. Hence, specifying the Helper can store all previous steps' state follows naturally from this.

Ah, this is the part I was missing, thanks for clarifying. I do think we should rope this into another PR though, since I think there could be more wording changes necessary to make this clear. For one we only call out recovering from `n` to `n+1`.

Reviewer 8: Opened a new PR on this edit: https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/569
We can resolve this thread here 😁

`collection-job-id` but with a different `CollectionReq` in the body MUST fail
with an HTTP client error status code.

After receiving the response to its `CollectionReq`, the Collector makes a

> request `GET /tasks/{tasks}/collection_jobs/{collection-job-id}` to check on

Reviewer 8: ```suggestion
request `GET /tasks/{task_id}/collection_jobs/{collection-job-id}` to check on
```
